### ISSUE TITLE
Provide a programmatic validation API

### ISSUE THREAD

**Issue by dandv** [ampproject/amphtml#1968]
Could https://cdn.ampproject.org/v0/validator.js be used to provide programmatic validation?

If all the validation logic is there, it would make things (#937, #1967), a lot easier if we could call an endpoint like

    https://cdn.ampproject.org/validate?url=...&callback=...

Automating the validation of lists of URLs would become a lot easier from anywhere that can make an HTTP request, such as Google Apps scripts that parse a spreadsheet of AMP URLs.

CC @pietergreyling, @jmadler, @adewale

**Comment by adewale** [comment]
@dandv please sketch out the API you'd like to see. What would be the output of this endpoint?

**Comment by jpettitt** [comment]
+1 for wanting an API.  

I'd like to see it accept both the URL of an AMP page, or the AMP html in a post body.

```
GET https://cdn.ampproject.org/v/?url= .... &callback=jsonpcallback
POST https://cdn.ampproject.org/v/  /*data in post body */
```
Result - something like

```
{
     "version": 0,   /* The amp version detected/validated */
     "source": "https://some.amp.site/page.html",   /* or "POST" if post data */
     "canonical": "http:// ...",  /* Canonical url from amp page */
     "valid":  false, /* or true */
     "extensions": [ /* array of amp extension detected */ ],   /* Nice to have */
     "errors": [
        {
              "position":  "10:2",
              "error": "The text (CDATA) inside tag 'author stylesheet' matches 'CSS !important', which is disallowed.",
              "help" : "https://github.com/ampproject/amphtml/blob/master/spec/amp-html-format.md#stylesheets"
        },
        {
              "position": "9:0",
              "warning": "AMP deprecated <style> boilerplate text detected",
              "help": "http:// ... "
        }
     ]
}
```
Nice to have: jsonp support,  optional html output.

**Comment by Gregable** [comment]
I think this might be a good idea, but nothing like that exists at the moment.

For now, it is possible to run the javascript validator in other contexts. I left a comment here:
https://github.com/ampproject/amphtml/issues/999#issuecomment-171787638 with an example.

You can also build it from source. There are instructions here:
https://github.com/ampproject/amphtml/tree/master/validator

**Comment by Gregable** [comment]
Following up, @dandv @jpettitt . What is the use case in mind here? It seems like using either the nodejs library (https://www.npmjs.com/package/amphtml-validator) or using validator.js directly are going to be faster in all cases than shipping an HTML document over the wire to cdn.ampproject.org and waiting for a response.

**Comment by dandv** [comment]
@Gregable: at the time, I was working on a piece of [Google Apps script](https://developers.google.com/apps-script/) attached to a trix with partner AMP URLs to test. In that environment, you can't import Node modules, so fetching the validation result over HTTP is the only realistic option.

The HTML document doesn't have to be shipped over; the URLs are publicly accessible, so the call would submit the URL for validation and receive a JSON result similar to what @jpettitt described.

**Comment by dknecht** [comment]
CloudFlare has added an endpoint to our beta cache with a json output.  We are happy to tweak output based of feedback. 

`curl https://cdn.edgeamp.org/q/' -X POST —data-binary @invalid_amp.html -H 'Content-Type: text/html; charset=UTF-8'`
`curl https://cdn.edgeamp.org/q/cfedgeorigin.com/amp/invalid_amp.html`



```
curl https://cdn.edgeamp.org/q/cfedgeorigin.com/amp/invalid_amp.html 2>/dev/null | python -mjson.tool
{
    "errors": [
        {
            "code": "MANDATORY_TAG_MISSING",
            "col": 0,
            "error": "The mandatory tag 'noscript enclosure for boilerplate' is missing or incorrect.",
            "help": "https://github.com/ampproject/amphtml/blob/master/spec/amp-boilerplate.md",
            "line": 12
        },
        {
            "code": "MANDATORY_TAG_MISSING",
            "col": 0,
            "error": "The mandatory tag 'head > style[amp-boilerplate]' is missing or incorrect.",
            "help": "https://github.com/ampproject/amphtml/blob/master/spec/amp-boilerplate.md",
            "line": 12
        },
        {
            "code": "MANDATORY_TAG_MISSING",
            "col": 0,
            "error": "The mandatory tag 'noscript > style[amp-boilerplate]' is missing or incorrect.",
            "help": "https://github.com/ampproject/amphtml/blob/master/spec/amp-boilerplate.md",
            "line": 12
        }
    ],
    "source": "http://cfedgeorigin.com/amp/invalid_amp.html",
    "valid": false,
    "version": "1471559432224"
}
```

**Comment by Gregable** [comment]
Looking into this. A few questions for those interested in this thread:

One of the main issues is avoiding abuse (ie: [DOS](https://en.wikipedia.org/wiki/Denial-of-service_attack)) via Google's fetching. We could solve this by:

1. Requiring that the API request provide the contents, rather than the URL, and we perform validation on the provided string. This wouldn't work for @dandv's case. Would this still be useful?
1. Only serving the cached validation result, which would be similar to just fetching the document via cdn.ampproject.org. This wouldn't support testing changes to a site, only seeing a snapshot in time. Would this still be useful?

**Comment by pietergreyling** [comment]
1) Requiring that the API request provide the contents, rather than the URL, and we perform validation on the provided string. This wouldn't work for @dandv's case. Would this still be useful?
- This could be useful especially in pre-production scenarios where the candidate AMP page is not yet hosted.

2) Only serving the cached validation result, which would be similar to just fetching the document via cdn.ampproject.org. This wouldn't support testing changes to a site, only seeing a snapshot in time. Would this still be useful?
- In a development scenario, which is iterative, this might not be so useful due to the time lag (unless there is a mechanism to trigger a refresh, which might bring us back full circle). 
- In a production troubleshooting use case, this could be pretty useful if I understand correctly.
  - This would imply caching validation results in FAIL cases as well as PASS regardless of whether the document was actually rejected (not cached) as in the FAIL case - correct? 
  - What would we store? The full validator output as at the time of validation? 
  - Would it make sense to also tag this with [the validator spec file revision](https://github.com/ampproject/amphtml/blob/master/validator/validator-main.protoascii#L28)?

**Comment by Gregable** [comment]
>

**Comment by Gregable** [comment]
>   - This would imply caching validation results in FAIL cases as well as PASS regardless of whether the document was actually rejected (not cached) as in the FAIL case - correct? 
>   - What would we store? The full validator output as at the time of validation? 

This is true, I don't think the cache currently stores FAIL cases, I think it retries on every request.

>   - Would it make sense to also tag this with [the validator spec file revision](https://github.com/ampproject/amphtml/blob/master/validator/validator-main.protoascii#L28)?

It wouldn't hurt, but I don't think get the feeling that most developers are using this revision number for much. It's really used only for development of the validator itself, we are quite careful about backwards incompatible changes these days.

**Comment by dandv** [comment]
> One of the main issues is avoiding abuse (ie: DOS) via Google's fetching.

Could DOS attacks via the "validate AMP URL" API be avoided via other anti-DOS mechanisms?

I.e. if we employ rate limiting, we can reasonably limit requests to a pretty low rate, given that:

* one AMP document is unlikely to realistically change more often than once per ~5 seconds, so validating more often than that wouldn't be useful
* multiple documents across a site tend to be generated from a small number of templates, so repeated sweeping validations across a domain can also be limited. API users would choose a representative set of AMP pages that exercise the templates.

If we implement domain whitelisting, we could have a verification mechanism similar to [verifying site ownership in Google Webmaster](https://support.google.com/webmasters/answer/35179?hl=en).

**Comment by brianlayman** [comment]
I got here as I was looking at a method would trigger a validation check when a git push occurs.  For now I'll probably implement the validator locally, but having an official URL even if it just returned pass/fail would be great!

**Comment by pietergreyling** [comment]
This could be of interest:
[https://github.com/ampproject/ampbench](https://github.com/ampproject/ampbench)
[Walkthrough article: Debug AMP pages with AMPBench, an open source app from the AMP Project.](https://medium.com/@greyling/ampbench-an-amp-url-validation-and-troubleshooting-application-d33ee83df604#)
Look for this section in the walkthrough article: "Experimental AMPBench JSON APIs"
The code is here:
[https://github.com/ampproject/ampbench/blob/master/ampbench_routes.js#L393](https://github.com/ampproject/ampbench/blob/master/ampbench_routes.js#L393)

**Comment by brianlayman** [comment]
@pietergreyling That looks to be perfect! Thank you for the supporting detail too. Time to start coding...

**Comment by ericlindley-g** [comment]
@pietergreyling  — this looks great! Are there plans to move the Experimental AMPBench JSON API to a stable status? We're seeing folks ask about this in the context of integrating into their platforms for faster in-context validation.

**Comment by pietergreyling** [comment]
@ericlindley-g At this point the recommended way of integrating AMPBench into a custom validation workflow using the JSON API is to run an instance of AMPBench on a dedicated hosting platform. The latter can be public or on an in-house server behind a firewall restricting access to internal clients. This also has the advantages of allowing the source code of the API to be tuned according to custom needs and the creation of Pull Requests to the AMPBench repository for any improvements based on such changes.

**Comment by ericlindley-g** [comment]
@pietergreyling Sounds good—thanks!

**Comment by johnnyshankman** [comment]
I'd still really appreciate this. There is currently no way to know if a page is valid using standard JS.

Can there not be a class selector on the HTML and/or an object accessible via window which holds the validation state when developers append `#development=1`?

This would be really useful for E2E testing via things like Robot Framework and Ghost Inspector.

**Comment by johnnyshankman** [comment]
@Gregable are there any new options for what I've described since this issue was opened? Looking for a reliable way to check validation status using just vanilla JS after appending `#development=1`
